src/util/ReferenceSet.ts

```diff
@@ -41,8 +41,11 @@ export default class ReferenceSet {
 
     let item,
       values = this.refs.values();
-    while (((item = values.next()), !item.done))
-      if (resolve(item.value) === value) return true;
+    while (((item = values.next()), !item.done)) {
+      const resolved = resolve(item.value);
+      if((Array.isArray(resolved) && resolved.some(r => r === value)) || resolved === value) return true;
```

Comment:
So this does work but it breaks the other case, where someone did "array().oneOf([ref('someArray')])` Admittedly that isn't very useful to do but it's confusing that array ref values are automatically flattened like this

Reply:
Interestingly, it does not break that case because of the `|| resolved == value`

Indeed, the following test passes:
```
    const r = [1,2,3];
    const m = array().oneOf([ref('$r')]).validateSync(r,{context:{r}});
    m.should.equal(r);
```

I could add it if you like.

But I also understand the hesitance to muddle this logic as well. It might seem more suitable to just instruct the user to add a custom test, which can easily be done in userland without further complicating the core.

Reply:
I like this feature...I wonder if maybe we just tag the ref with `flatten = true` or something and then check for that in the ref set resolution logic?

Reply:
@jquense well, to be clear, due to #1437 getting merged, you can no longer do an equality comparison on the arrays - i.e. array().oneOf([ref('$someArray')]) - would no longer work due to the way concat works [here](https://github.com/jquense/yup/blob/master/src/util/ReferenceSet.ts#L30). 

Maybe I can fix that behaviour (allow the equality check) in this PR, and add the flatten flag as you suggest?